packages/swing-store/src/bundleStore.js

```
@@ -62,20 +66,36 @@ export function makeBundleStore(db, ensureTxn, noteExport = () => {}) {
     return `b${Nat(version)}-${hash}`;
   }
 
-  const sqlAddBundle = db.prepare(`
-    INSERT OR REPLACE INTO bundles
-      (bundleID, bundle)
-    VALUES (?, ?)
+  // the PRIMARY KEY constraint requires the bundleID not already
+  // exist
+  const sqlAddBundleRecord = db.prepare(`
+INSERT INTO bundles (bundleID, bundle) VALUES (?, NULL)
   `);
 
-  /**
-   * Store a bundle.  Here the bundle itself is presumed valid.
-   *
-   * @param {string} bundleID
-   * @param {Bundle} bundle
-   */
-  function addBundle(bundleID, bundle) {
+  // this sees both populated and pruned (not-yet-populated) records
+  const sqlHasBundleRecord = db.prepare(`
+    SELECT count(*)
+    FROM bundles
+    WHERE bundleID = ?
+  `);
+  sqlHasBundleRecord.pluck();
+
+  const sqlPopulateBundleRecord = db.prepare(`
+    UPDATE bundles SET bundle = ? WHERE bundleID = ?
+  `);
+
+  function addBundleRecord(bundleID) {
+    ensureTxn();
+    sqlAddBundleRecord.run(bundleID);
+  }
+
+  function populateBundle(bundleID, serialized) {
     ensureTxn();
+    sqlHasBundleRecord.run(bundleID) || Fail`missing ${bundleID}`;
```

I supposed we're not deferring the existence check to the `UPDATE` that would fail if the record doesn't exist in order to provide a better error message? That's fine, it's likely a cheap enough check to make.

---

In SQL, `UPDATE bundles SET bundle = ? WHERE bundleID = ?` will silently do nothing when WHERE constraint doesn't match any rows. (Just like it will silently corrupt your entire table if you forget the WHERE clause). So the explicit check is actually load-bearing.

---

Ugh, update silently doing nothing is somewhat surprising...